File: src/main/java/com/github/nsc/de/shake/parser/node/objects/ConstructorDeclarationNode.java
package com.github.nsc.de.shake.parser.node.objects;

import com.github.nsc.de.shake.lexer.characterinput.position.PositionMap;
import com.github.nsc.de.shake.parser.node.AccessDescriber;
import com.github.nsc.de.shake.parser.node.Tree;
import com.github.nsc.de.shake.parser.node.ValuedNode;
import com.github.nsc.de.shake.parser.node.functions.FunctionArgumentNode;

import java.util.Arrays;

public class ConstructorDeclarationNode extends ValuedNode {

    private final String name;
    private final Tree body;
    private final FunctionArgumentNode[] args;
    private final AccessDescriber access;

    public ConstructorDeclarationNode(PositionMap map, String name, Tree body, FunctionArgumentNode[] args) {
        this(map, name, body, args, AccessDescriber.PACKAGE);
    }


    public ConstructorDeclarationNode(PositionMap map, String name, Tree body, FunctionArgumentNode[] args,
                                      AccessDescriber access) {
        super(map);
        this.name = name;
        this.body = body;
        this.args = args;
        this.access = access;
    }


    public ConstructorDeclarationNode(PositionMap map, Tree body, FunctionArgumentNode[] args) {
        this(map, body, args, AccessDescriber.PACKAGE);
    }


    public ConstructorDeclarationNode(PositionMap map, Tree body, FunctionArgumentNode[] args, AccessDescriber access) {
        this(map, null, body, args, access);
    }

    public String getName() {
        return name;
    }
    public Tree getBody() {
        return body;
    }
    public FunctionArgumentNode[] getArgs() {
        return args;
    }
    public AccessDescriber getAccess() { return access; }

    @Override
    public String toString() {
        return "FunctionDeclarationNode{" +
                "name='" + name + '\'' +
                ", body=" + body +
                ", args=" + Arrays.toString(args) +
                '}';
    }
}

File: parser/src/main/kotlin/com/github/nsc/de/shake/parser/node/objects/ConstructorDeclarationNode.kt
package com.github.nsc.de.shake.parser.node.objects

import com.github.nsc.de.shake.parser.node.ValuedNode
import com.github.nsc.de.shake.parser.node.AccessDescriber
import com.github.nsc.de.shake.parser.node.Tree
import com.github.nsc.de.shake.parser.node.functions.FunctionArgumentNode
import com.github.nsc.de.shake.util.characterinput.position.PositionMap

@Suppress("unused")
class ConstructorDeclarationNode
@JvmOverloads constructor(

    map: PositionMap,
    val name: String?,
    val body: Tree,
    val args: Array<FunctionArgumentNode>,
    val access: AccessDescriber? = AccessDescriber.PACKAGE

) : ValuedNode(map) {

    @JvmOverloads
    constructor(
        map: PositionMap,
        body: Tree,
        args: Array<FunctionArgumentNode>,
        access: AccessDescriber? = AccessDescriber.PACKAGE
    ) : this(map, null, body, args, access)

    override fun toJson(): Map<String, *> =
        mapOf(
            "name" to "ConstructorDeclarationNode",
            "function_name" to name,
            "args" to args.map { it.json },
            "body" to body.json,
            "access" to access.toString(),
        )
}